### PR TITLE
MOM6 Angle Calculation

### LINKED ISSUE #31: Change the Angle Calculation used in the inputted Hgrid to the MOM6 Version
Following up on Fred's comment, the angle calculation of the Hgrid in MOM is not read as part of the file and is instead calculated in MOM. This issue is about calculating the hgrid angle the same way as MOM in regional mom6

### PR DESCRIPTION
**Pull request by manishvenu**
Fixes #31 

Implement the MOM6 Angle Calculation in Regional MOM6. Add functions, classes to RM6 and adjust the regridding functions to accommodate these new angles. 

[Explanation/Visualization](https://docs.google.com/presentation/d/1jNEDsFAMclZ9Yh770IWdOZRLqrK1elTaO43KTtlICKk/edit#slide=id.g318d8e6edd6_0_5) of the changes.

Major Notes & Changes:

Please see the document `angle_calc.md` for a more complete explanation
1. New file (`rotation.py`) to implement MOM6 angle calculation & Fred, Keiths methods:
    1. MOM6 `modulo_around_point`: Implement MOM6 modulo_around_point. 
    3. Angle Calculation from MOM6 `mom6_angle_calculation_method`:  Implement MOM6 intialize_grid_rotation_angle function logic. Only the new version from MOM6 is included
    4.  `initialize_grid_rotation_angle`: This calculates t-point angle, which is what MOM6 does using the mom6_angle_calculation_method
    5. `initialize_grid_rotation_angles_using_pseudo_hgrid`: Using Fred's pseudo hgrid that has an extra boundary row & col, calculate angle at every point in the original hgrid
    6. `create_pseudo_hgrid`: Fred calculation creating an extra boundary row & col on hgrid
    7. `RotationMethod` Enum Class: This class is to determine by which method we rotate the boundary conditions.
 1.  Adjustments to `regridding.py`:
       1. `get_hgrid_arakawa_c_points`: Work from Fred splitting hgrid into the t, u, v, & q points converted to a RM6 function
       1. `coords` changes: A change to the `coords` function to get the boundary t-points for Keith's method.
 1. Adjustments to `regrid_tides` & `regrid_velocity_tracers`:
       1.  What: accommodate the different rotation methods using the Enum Class (Keith, Fred, Old)
       2. Code: uses if statements
       3. Visualization: The [Visualization](https://docs.google.com/presentation/d/1jNEDsFAMclZ9Yh770IWdOZRLqrK1elTaO43KTtlICKk/edit#slide=id.g318d8e6edd6_0_5) is a great explanation of this
 1. Adjustments to testing:
       1. `test_rotation.py`:  New test file for the rotation
       4. **SHORT TERM CHANGE** Docker image adjustment: Updated the docker image we use with RM6 github testing with a curvilinear grid for use in `test_rotation.py`, that is a crocodile image until someone with permissions updates the RM6 docker image, maybe we should have a CROCODILE one to? But probably not
       5. `conftest.py`: Adds a fixture to skip a test if the files on the docker image don't exist on the computer (this is good for local testing)
       6. Minor test cleaning: A bunch of small changes to a few tests to not create extra folders.
 7. Doc Changes: 
      1. Angle Calc: `angle_calc.md` explaining this angle issue, and 
      1. Docker: `docker_image_dev.md` for changing the docker image in the future.
 10. Visualizations (Notebook to look at the differences in the methods: `/glade/u/home/manishrv/documents/nwa12_0.1/tides_dev/test_rotational_methods.ipynb`)

### CODE REVIEW

**Comment by manishvenu** [comment]
Quick note: The docker image was interesting...and also was not at all used in RM6 testing I'm pretty sure. There is a ton of data still in it. If we like how the testing is setup currently, I'll go ahead and delete all that & keep the image small -> probably only a few MB. Or make a whole new image...

File: regional_mom6/regional_mom6.py [hunk 1]
```diff
@@ -3310,8 +3494,15 @@ def regrid_tides(
         self.encode_tidal_files_and_output(ds_ap, "tz")
 
         ########### Regrid Tidal Velocity ######################
-        regrid_u = rgd.create_regridder(tpxo_u[["lon", "lat", "uRe"]], coords, ".temp")
-        regrid_v = rgd.create_regridder(tpxo_v[["lon", "lat", "vRe"]], coords, ".temp2")
+
+        # Change to t-point coords
+        if rotational_method == rot.RotationMethod.KEITH_DOUBLE_REGRIDDING:
```
Comment (alperaltuntas) [inline regional_mom6/regional_mom6.py:3499]: I wonder if it's worth keeping KEITH_DOUBLE_REGRIDDING in the code. Unless there is a use case of this method over Fred's method, I suggest we remove it to keep the code simpler.

**Review by alperaltuntas — APPROVED** [review]

File: regional_mom6/regional_mom6.py (hunk 1 above)
Comment (manishvenu) [reply regional_mom6/regional_mom6.py:3499]: Removing KEITH_DOUBLE_REGRIDDING!

File: regional_mom6/rotation.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+import numpy as np
+from .regridding import get_hgrid_arakawa_c_points
+
+
+class RotationMethod(Enum):
+    """
+    This Enum defines the rotational method to be used in boundary conditions. The main regional mom6 class passes in this enum to regrid_tides and regrid_velocity_tracers to determine the method used.
+
+    KEITH_DOUBLE_REGRIDDING: This method is used to regrid the boundary conditions to the t-points, b/c we can calculate t-point angle the same way as MOM6, rotate the conditions, and regrid again to the q-u-v, or actual, boundary
+    FRED_AVERAGE: This method is used with the basis that we can find the angles at the q-u-v points by pretending we have another row/column of the hgrid with the same distances as the t-point to u/v points in the actual grid then use the four poitns to calculate the angle the exact same way MOM6 does.
```
[the remark this answers is not in the document]
Comment (manishvenu) [reply regional_mom6/rotation.py:16]: Changed to EXPAND_GRID - happy to take criticism on this name

File: regional_mom6/rotation.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    GIVEN_ANGLE: This is the original default RM6 method which expects a pre-given angle called angle_dx
+    NO_ROTATION: Grids parallel to lat/lon axes, no rotation needed
```
[the remark this answers is not in the document]
Comment (manishvenu) [reply regional_mom6/rotation.py:17]: Fixed with a utils function is_rectilinear_hgrid

**Comment by manishvenu** [comment]
Checks passed, so I'm merging. If there's any last comments on this, they can be addressed in #33

**Pull request merged**